velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp

```diff
@@ -13,89 +13,214 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 #include "velox/exec/Aggregate.h"
 #include "velox/expression/FunctionSignature.h"
 #include "velox/functions/prestosql/aggregates/AggregateNames.h"
+#include "velox/functions/prestosql/aggregates/SingleValueAccumulator.h"
 #include "velox/vector/ComplexVector.h"
 #include "velox/vector/DecodedVector.h"
 #include "velox/vector/FlatVector.h"
-
 namespace facebook::velox::aggregate {
 
 namespace {
 
-// MinMaxByAggregate is the base class for min_by and max_by functions. These
-// functions return the value of X associated with the minimum/maximum value of
-// Y over all input values. Partial aggregation produces a pair of X and min/max
-// Y. Final aggregation takes a pair of X and min/max Y and returns X. T is the
-// type of X and U is the type of Y.
+template <typename T>
+constexpr bool isNumeric() {
+  return std::is_same_v<T, int8_t> || std::is_same_v<T, int16_t> ||
+      std::is_same_v<T, int32_t> || std::is_same_v<T, int64_t> ||
+      std::is_same_v<T, float> || std::is_same_v<T, double>;
+}
+
+template <typename T, typename TAccumulator>
+void extract(
+    TAccumulator* accumulator,
+    const VectorPtr& vector,
+    vector_size_t index,
+    T* rawValues) {
+  if constexpr (isNumeric<T>()) {
+    rawValues[index] = *accumulator;
+  } else {
+    VELOX_DCHECK(accumulator->hasValue());
+    accumulator->read(vector, index);
+  }
+}
+
+template <typename T, typename TAccumulator>
+void store(
+    TAccumulator* accumulator,
+    const DecodedVector& decodedVector,
+    vector_size_t index,
+    HashStringAllocator* allocator) {
+  if constexpr (isNumeric<T>()) {
+    *accumulator = decodedVector.valueAt<T>(index);
+  } else {
+    accumulator->write(
+        decodedVector.base(), decodedVector.index(index), allocator);
+  }
+}
+
+template <typename T, typename TAccumulator>
+bool greaterThan(
```

Comment:
Method-level comment and a comment inside the body of the function could be useful. It is confusing to the reader that one branch returns the result of > , but the other < 0.

```
/// Returns true if the value in 'index' row of 'newComparisons' is strictly greater than the value in the 'accumulator'.
...

// SingleValueAccumulator::compare has the semantics of accumulator-value-is-less-than-vector-value.
return !accumulator->hasValue() ||
        (accumulator->compare(newComparisons, index) < 0);
```